File: solutions_2453486_0/Python/theceem/tttt.py
#!/usr/bin/env python
import sys

def read_test_case(f):
  # read board and following empty line
  board = [f.readline().strip() for i in range(4)]
  assert not f.readline().strip(), "Each test case is followed by an empty line."
  return board

def analyze_quad(s):
  for player in ['X', 'O']:
    if s.count(player) == 4 or (s.count(player) == 3 and s.count('T') == 1):
      return '%s won' % player

def analyze(board):
  # "if there is a row, column or a diagonal containing 4 of that player's symbols,
  #  or containing 3 of her symbols and the 'T' symbol,
  #  she wins and the game ends"

  # collect quads to test
  quads = list(board)   # rows
  for i in range(4):
    quads.append([board[j][i] for j in range(4)])   # columns
  quads.append([board[i][i] for i in range(4)])     # principal diagonal
  quads.append([board[i][3-i] for i in range(4)])   # other diagonal

  # test each quad
  for quad in quads:
    result = analyze_quad(quad)
    if result:
      return result

  # "if all of the fields are filled with symbols and nobody won, the game ends in a draw"
  if not '.' in ''.join(board):
    return 'Draw'

  return 'Game has not completed'

# process all tests
test_count = int(sys.stdin.readline())
for i in range(test_count):
  board = read_test_case(sys.stdin)
  print 'Case #%d: %s' % (i+1, analyze(board))

File: solutions_2453486_1/C++/RoastDuck/tttt.cpp
#include<cstdio>
#include<cstring>
#include<algorithm>
using namespace std;

int T;
char map[7][7];

bool Is(char x, char c)
{
	return (x==c || x=='T');
}

bool win(char c)
{
	if (Is(map[1][1],c) && Is(map[1][2],c) && Is(map[1][3],c) && Is(map[1][4],c)) return true;
	if (Is(map[2][1],c) && Is(map[2][2],c) && Is(map[2][3],c) && Is(map[2][4],c)) return true;
	if (Is(map[3][1],c) && Is(map[3][2],c) && Is(map[3][3],c) && Is(map[3][4],c)) return true;
	if (Is(map[4][1],c) && Is(map[4][2],c) && Is(map[4][3],c) && Is(map[4][4],c)) return true;
	
	if (Is(map[1][1],c) && Is(map[2][1],c) && Is(map[3][1],c) && Is(map[4][1],c)) return true;
	if (Is(map[1][2],c) && Is(map[2][2],c) && Is(map[3][2],c) && Is(map[4][2],c)) return true;
	if (Is(map[1][3],c) && Is(map[2][3],c) && Is(map[3][3],c) && Is(map[4][3],c)) return true;
	if (Is(map[1][4],c) && Is(map[2][4],c) && Is(map[3][4],c) && Is(map[4][4],c)) return true;
	
	if (Is(map[1][1],c) && Is(map[2][2],c) && Is(map[3][3],c) && Is(map[4][4],c)) return true;
	if (Is(map[1][4],c) && Is(map[2][3],c) && Is(map[3][2],c) && Is(map[4][1],c)) return true;
	
	return false;
}

int main()
{
	freopen("tttt.in","r",stdin);
	freopen("tttt.out","w",stdout);
	scanf("%d",&T);
	for (int t=1;t<=T;t++)
	{
		printf("Case #%d: ",t);
		for (int i=1;i<=4;i++) scanf("%s",map[i]+1);
		if (win('X'))
		{
			printf("X won\n");
			continue;
		}
		if (win('O'))
		{
			printf("O won\n");
			continue;
		}
		
		bool draw(true);
		for (int i=1;i<=4;i++)
			for (int j=1;j<=4;j++)
				if (map[i][j]=='.')
				{
					draw=false;
					break;
				}
		if (draw)
			printf("Draw\n");
		else
			printf("Game has not completed\n");
	}
	return 0;
}

